### ISSUE TITLE
filter_chain_match not matching on ip address

### ISSUE THREAD

**Issue by grayaii** [envoyproxy/envoy#8396]
*Title*: *filter_chain_match not matching on source_prefix_ranges*

*Description*:
I have 2 filter_chain_match blocks.
One to catch all requests coming from 172.X.Y.Z and another to catch the rest, 0.0.0.0

It seems indeterministic which filter chain is used.  When I deploy envoy, sometime it uses one filter chain, and other times, it uses the other.

Maybe I'm getting source_prefix_ranges vs prefix_ranges confused, but I tried both, and none seem to work.

**Bug Template**

    admin:
      access_log_path: /dev/null
      address:
        socket_address:
          protocol: TCP
          address: 127.0.0.1
          port_value: 8999

    node:
      cluster: cco-envoy-cluster
      id: cco-envoy-id

    static_resources:
      listeners:
      - name: listener_0
        address:
          socket_address:
            address: 0.0.0.0
            port_value: 8080

        filter_chains:
        # If source IP is 172.X.X.X, use this:
        - filter_chain_match:
          source_prefix_ranges:
          - address_prefix: 172.0.0.0
            prefix_len: 8

          filters:
          - name: envoy.http_connection_manager
            config:
              server_name: "cco-router-stats"
              http_protocol_options:
                accept_http_10: true
              stat_prefix: ingress_http
              codec_type: AUTO
              route_config:
                name: local_route
                virtual_hosts:
                - name: envoy_stats
                  domains:
                    - "*"
                  routes:
                    - match:
                        prefix: /stats
                      route:
                        cluster: envoy_stats
              http_filters:
              - name: envoy.router
                config:
                  suppress_envoy_headers: true

        # For all other Source IP adddresses, use this:
        - filter_chain_match:
          source_prefix_ranges:
          - address_prefix: 0.0.0.0
            prefix_len: 0

          filters:
          - name: envoy.http_connection_manager
            config:
              server_name: "cco-router"
              http_protocol_options:
                accept_http_10: true
              stat_prefix: ingress_http
              codec_type: AUTO
              route_config:
                name: local_route
                virtual_hosts:
                - name: heath_checker
                  domains:
                  - "*"
                  routes:
                  - match: { prefix: "/health_check" }
                    direct_response:
                      status: 200
                      body:
                        inline_string: ok
              
              http_filters:
              - name: envoy.router
                config:
                  suppress_envoy_headers: true

      clusters:
      - name: envoy_stats
        connect_timeout: 0.250s
        type: LOGICAL_DNS
        lb_policy: ROUND_ROBIN
        hosts:
          - socket_address:
              protocol: TCP
              address: 127.0.0.1
              port_value: 8999

*Repro steps*:
When I curl localhost/stats ONLY from a 172.x.y.z address, I would expect it to work.
When I curl localhost/health_check ONLY from non-172.x.y.z address, I would expect it to work.

**Comment by alyssawilk** [comment]
Yeah, I thought Envoy did top down for route matching so I would expect deterministic behavior here.
Could you post at trace log of one of your surprising match failures?  That might help with debug

**Comment by grayaii** [comment]
Sorry for the delay. I'm still trying to get that information for you.
There must be something weird with the exact config that I posted (must be an indent issue).
When I convert that that yaml that I pasted to json, I can see an issue:
<img width="639" alt="typo" src="https://user-images.githubusercontent.com/847072/65720244-8ba2ab80-e075-11e9-983b-366f33b1306c.png">

Once I get that sorted out, I'll paste the trace.
